source/extensions/filters/network/zookeeper_proxy/filter.cc

```diff
@@ -19,55 +21,169 @@ namespace NetworkFilters {
+  if (latency <= latency_threshold) {
+    return ErrorBudgetResponseType::FAST;
+  }
+
+  return ErrorBudgetResponseType::SLOW;
+}
+
+void ZooKeeperFilterConfig::initOpCode(OpCodes opcode, Stats::Counter& resp_counter,
+                                       Stats::Counter& resp_fast_counter,
+                                       Stats::Counter& resp_slow_counter, absl::string_view name) {
   OpCodeInfo& opcode_info = op_code_map_[opcode];
-  opcode_info.counter_ = &counter;
+  opcode_info.resp_counter_ = &resp_counter;
+  opcode_info.resp_fast_counter_ = &resp_fast_counter;
+  opcode_info.resp_slow_counter_ = &resp_slow_counter;
   opcode_info.opname_ = std::string(name);
   opcode_info.latency_name_ = stat_name_set_->add(absl::StrCat(name, "_latency"));
 }
 
+int32_t ZooKeeperFilterConfig::getOpCodeIndex(LatencyThreshold_Opcode opcode) {
+  static absl::flat_hash_map<LatencyThreshold_Opcode, int32_t> opcode_map = {
+      // Use a special key (-999) to store the default latency threshold.
+      {LatencyThreshold::Default, -999},
+      {LatencyThreshold::Connect, 0},
+      {LatencyThreshold::Create, 1},
+      {LatencyThreshold::Delete, 2},
+      {LatencyThreshold::Exists, 3},
+      {LatencyThreshold::GetData, 4},
+      {LatencyThreshold::SetData, 5},
+      {LatencyThreshold::GetAcl, 6},
+      {LatencyThreshold::SetAcl, 7},
+      {LatencyThreshold::GetChildren, 8},
+      {LatencyThreshold::Sync, 9},
+      {LatencyThreshold::Ping, 11},
+      {LatencyThreshold::GetChildren2, 12},
+      {LatencyThreshold::Check, 13},
+      {LatencyThreshold::Multi, 14},
+      {LatencyThreshold::Create2, 15},
+      {LatencyThreshold::Reconfig, 16},
+      {LatencyThreshold::CheckWatches, 17},
+      {LatencyThreshold::RemoveWatches, 18},
+      {LatencyThreshold::CreateContainer, 19},
+      {LatencyThreshold::CreateTtl, 21},
+      {LatencyThreshold::Close, -11},
+      {LatencyThreshold::SetAuth, 100},
+      {LatencyThreshold::SetWatches, 101},
+      {LatencyThreshold::GetEphemerals, 103},
+      {LatencyThreshold::GetAllChildrenNumber, 104},
+      {LatencyThreshold::SetWatches2, 105}};
+
+  ASSERT(opcode_map.find(opcode) != opcode_map.end());
+  return opcode_map[opcode];
+}
+
+absl::flat_hash_map<int32_t, std::chrono::milliseconds>
+ZooKeeperFilterConfig::parseLatencyThresholds(const LatencyThresholdList& latency_thresholds) {
+  absl::flat_hash_map<int32_t, std::chrono::milliseconds> latency_threshold_map;
+  for (const auto& threshold : latency_thresholds) {
+    latency_threshold_map[getOpCodeIndex(threshold.opcode())] =
+        std::chrono::milliseconds(PROTOBUF_GET_MS_REQUIRED(threshold, threshold));
+  }
+  return latency_threshold_map;
+}
+
+std::chrono::milliseconds ZooKeeperFilterConfig::getDefaultLatencyThreshold() {
+  if (latency_threshold_map_.empty()) {
+    return std::chrono::milliseconds(0);
+  }
+
+  std::chrono::milliseconds default_latency_threshold(100);
+  auto it = latency_threshold_map_.find(-999);
```

Comment:
This "magic" number should either be removed (maybe by using an explicit default value field), or it should be properly documented here.

Reply:
Yeah, I documented [here](https://github.com/envoyproxy/envoy/pull/26594/files#diff-a4f1b5b91a56b1929b870a98140d5ca6fe188a446c1acb64e42bc0fbdcbc5a94R131), which is a mapping between proto value and opcode value.